src/lib/utils.rs

```diff
@@ -187,6 +198,187 @@ To compress an uncompressed FASTQ file with bgzip:
     Err(anyhow!(message).context(context))
 }
 
+pub static INPUT_FASTQ_SUFFIX: &str = "_001.fastq.gz";
+
+lazy_static! {
+    /// <*>_L00#_<R# or I#>_001.fastq.gz
+    static ref INPUT_FASTQ_REGEX: Regex = Regex::new(r"^(.*)_L00(\d{1})_([RIUS])(\d{1})_001.fastq.gz$").unwrap();
+}
+
+/// Contains information about a FASTQ that has been inferred from the file name.  This includes
+/// lane, segment type (e.g. template/read, sample barcode/index, molecular barcode/index, and
+/// skip bases), and segment number (e.g. R1 or R2).
+#[derive(Debug, Clone)]
+pub struct InputFastq {
+    pub path: PathBuf,
+    pub prefix: String,
+    pub lane: usize,
+    pub kind: SegmentType,
+    pub kind_number: i32,
+}
+
+impl InputFastq {
+    /// Create a new `InputFastq` inferring information from the file name.  This must match the
+    /// `INPUT_FASTQ_REGEX` pattern.
+    pub fn new<P: AsRef<Path>>(path: P) -> Option<InputFastq> {
+        let file_name = path.as_ref().file_name().unwrap().to_string_lossy();
+        match INPUT_FASTQ_REGEX.captures(&file_name) {
+            None => None,
+            Some(captures) => {
+                let prefix = captures.get(1).unwrap().as_str().to_string();
+                let lane = captures.get(2).unwrap().as_str().parse::<usize>().unwrap();
+                let kind: SegmentType = match captures.get(3).unwrap().as_str() {
+                    "R" => SegmentType::Template,
+                    "I" => SegmentType::SampleBarcode,
+                    "U" => SegmentType::MolecularBarcode,
+                    "S" => SegmentType::Skip,
+                    knd => panic!("Could not determine kind from {}", knd),
+                };
+                let kind_number = captures.get(4).unwrap().as_str().parse::<i32>().unwrap();
+
+                Some(InputFastq {
+                    path: path.as_ref().to_path_buf(),
+                    prefix,
+                    lane,
+                    kind,
+                    kind_number,
+                })
+            }
+        }
+    }
+
+    /// Identifies all FASTQs that share the common path prefix and match the
+    /// `INPUT_FASTQ_REGEX` pattern.  The path prefix may also be a directory.  The FASTQS are
+    /// returned in sorted order.
+    pub fn slurp<P: AsRef<Path>>(path_prefix: P) -> Vec<InputFastq> {
+        let (parent, prefix) = if path_prefix.as_ref().is_file() {
+            (
+                path_prefix.as_ref().parent().unwrap(),
+                path_prefix.as_ref().file_name().unwrap().to_str().unwrap(),
```

Comment:
This doesn't make a lot of sense to me.  `is_file()` should only return true for an extant file - I think this function only works if you pass in a directory with no other prefix as a result.

Also, if you inverted this (using `is_dir()`) these unwraps can very much fail.  E.g. with `--fastqs Undetermined` while the _file_ has a parent, the _path_ doesn't.

Reply:
Take one more look